File: .github/scripts/detect-changes/src/lib/services.ts
/**
 * Service discovery module for docker-compose.yml parsing.
 *
 * This module provides functionality to discover and extract metadata from services
 * defined in docker-compose.yml files. It uses `docker compose config` to parse
 * the compose file with environment variable substitution and extracts service
 * information including build context, dockerfile paths, and build arguments.
 */

import { execFileSync } from 'child_process';
import type { Service } from './types.js';

/**
 * Represents the structure of docker-compose config output.
 */
interface ComposeConfig {
  services: Record<string, ServiceConfig>;
}

/**
 * Represents a service configuration from docker-compose.yml.
 */
interface ServiceConfig {
  image?: string;
  build?: string | BuildConfig;
}

/**
 * Represents build configuration in docker-compose.yml.
 */
interface BuildConfig {
  context: string;
  dockerfile?: string | null;
  args?: Record<string, string> | null;
}

/**
 * Discovers services from docker-compose.yml using `docker compose config`.
 *
 * This function executes `docker compose config --format json` to get the parsed
 * compose configuration with all environment variable substitutions applied.
 * It then extracts metadata for all services that have a build directive.
 *
 * @param composeFile Path to docker-compose.yml file
 * @param envFile Path to .env file for environment variable substitution
 * @returns Array of discovered services with their metadata
 * @throws Error if docker compose command fails or returns invalid JSON
 *
 * @example
 * ```typescript
 * const services = discoverServicesFromCompose('docker-compose.yml', '.env');
 * console.log(services[0].service_name); // 'broker'
 * console.log(services[0].build_context); // 'docker/mosquitto'
 * ```
 */
export function discoverServicesFromCompose(composeFile: string, envFile: string): Service[] {
  try {
    // Execute docker compose config and capture JSON output
    const output = execFileSync(
      'docker',
      ['compose', '--env-file', envFile, '--file', composeFile, 'config', '--format', 'json'],
      {
        encoding: 'utf-8',
        maxBuffer: 10 * 1024 * 1024, // 10MB buffer for large compose files
      }
    );

    // Parse JSON output
    const config: ComposeConfig = JSON.parse(output);

    // Extract metadata for all services
    const services: Service[] = [];
    for (const [serviceName, serviceConfig] of Object.entries(config.services || {})) {
      const metadata = extractServiceMetadata(serviceName, serviceConfig);
      if (metadata !== null) {
        services.push(metadata);
      }
    }

    return services;
  } catch (error) {
    // Re-throw with context for better error messages
    if (error instanceof Error) {
      throw error;
    }
    throw new Error(`Failed to discover services: ${String(error)}`);
  }
}

/**
 * Extracts service metadata from docker-compose service configuration.
 *
 * This function parses the service configuration object and extracts relevant
 * metadata including service name, image, build context, dockerfile path, and
 * build arguments. It handles both short build syntax (build: "path") and
 * long build syntax (build: { context, dockerfile, args }).
 *
 * @param serviceName Name of the service in docker-compose.yml
 * @param serviceConfig Service configuration object from docker-compose
 * @returns Service metadata object, or null if service has no build directive
 *
 * @example
 * ```typescript
 * const config = {
 *   image: 'ghcr.io/groupsky/homy/mosquitto:latest',
 *   build: { context: 'docker/mosquitto' }
 * };
 * const metadata = extractServiceMetadata('broker', config);
 * console.log(metadata.dockerfile_path); // 'docker/mosquitto/Dockerfile'
 * ```
 */
export function extractServiceMetadata(
  serviceName: string,
  serviceConfig: ServiceConfig
): Service | null {
  // Skip services without build directive
  if (!serviceConfig.build) {
    return null;
  }

  let buildContext: string;
  let dockerfilePath: string;
  let buildArgs: Record<string, string> | undefined;

  // Handle short build syntax: build: "docker/path"
  if (typeof serviceConfig.build === 'string') {
    buildContext = serviceConfig.build;
    dockerfilePath = normalizePath(buildContext, 'Dockerfile');
    buildArgs = undefined;
  } else {
    // Handle long build syntax: build: { context, dockerfile, args }
    buildContext = serviceConfig.build.context;

    // Resolve dockerfile path
    const dockerfile = serviceConfig.build.dockerfile || 'Dockerfile';
    dockerfilePath = normalizePath(buildContext, dockerfile);

    // Extract build args (filter out empty objects and null)
    const args = serviceConfig.build.args;
    if (args && typeof args === 'object' && Object.keys(args).length > 0) {
      buildArgs = args;
    } else {
      buildArgs = undefined;
    }
  }

  return {
    service_name: serviceName,
    image: serviceConfig.image,
    build_context: buildContext,
    dockerfile_path: dockerfilePath,
    build_args: buildArgs,
  };
}

/**
 * Filters services to only those using GHCR base images or with build directives.
 *
 * This function filters the services array to include only:
 * 1. Services with images starting with 'ghcr.io/groupsky/homy/'
 * 2. Services with images starting with 'ghcr.io/home-assistant/' (allowed exception)
 * 3. Services without an image field but with a build directive (local builds)
 *
 * @param services Array of services to filter
 * @returns Filtered array containing only GHCR-based services
 *
 * @example
 * ```typescript
 * const services = [
 *   { service_name: 'broker', image: 'ghcr.io/groupsky/homy/mosquitto:latest', ... },
 *   { service_name: 'postgres', image: 'postgres:15', ... }
 * ];
 * const filtered = filterGhcrServices(services);
 * console.log(filtered.length); // 1 (only broker)
 * ```
 */
export function filterGhcrServices(services: Service[]): Service[] {
  return services.filter((service) => {
    // Include services without image field (local builds with build directive)
    if (!service.image) {
      return true;
    }

    // Include services using GHCR base images
    if (service.image.startsWith('ghcr.io/groupsky/homy/')) {
      return true;
    }

    // Include Home Assistant images (allowed exception)
    if (service.image.startsWith('ghcr.io/home-assistant/')) {
      return true;
    }

    // Exclude all other images
    return false;
  });
}

/**
 * Builds a mapping from service names (including aliases) to their canonical service names.
 *
 * Service aliases are docker-compose services that use the same Docker image as another
 * service but with different runtime configuration. They should resolve to the canonical
 * service name for build purposes.
 *
 * A canonical service is one where the service name matches the image name.
 * An alias is any other service using that same image.
 *
 * @param services Array of all services
 * @returns Map from service name to canonical service name
 *
 * @example
 * ```typescript
 * const services = [
 *   { service_name: 'historian', image: 'ghcr.io/groupsky/homy/historian:latest', ... },
 *   { service_name: 'historian-secondary', image: 'ghcr.io/groupsky/homy/historian:latest', ... }
 * ];
 * const mapping = buildServiceAliasMapping(services);
 * console.log(mapping.get('historian')); // 'historian'
 * console.log(mapping.get('historian-secondary')); // 'historian'
 * ```
 */
export function buildServiceAliasMapping(services: Service[]): Map<string, string> {
  // First pass: find canonical services (service name matches image name)
  const imageToCanonicalService = new Map<string, string>();

  for (const service of services) {
    if (!service.image) {
      // Services without images (local builds) map to themselves
      imageToCanonicalService.set(service.service_name, service.service_name);
      continue;
    }

    // Extract image name without tag (e.g., "ghcr.io/groupsky/homy/historian:latest" -> "historian")
    const imageParts = service.image.split('/');
    const imageNameWithTag = imageParts[imageParts.length - 1];
    const imageName = imageNameWithTag.split(':')[0];

    // If service name matches image name, it's canonical
    if (service.service_name === imageName) {
      imageToCanonicalService.set(service.image, service.service_name);
    }
  }

  // Second pass: build alias mapping
  const aliasMapping = new Map<string, string>();

  for (const service of services) {
    if (!service.image) {
      // Local builds map to themselves
      aliasMapping.set(service.service_name, service.service_name);
      continue;
    }

    // Look up canonical service for this image
    const canonicalService = imageToCanonicalService.get(service.image);

    if (canonicalService) {
      // Map to canonical service
      aliasMapping.set(service.service_name, canonicalService);
    } else {
      // No canonical service found - map to itself (orphan image)
      aliasMapping.set(service.service_name, service.service_name);
    }
  }

  return aliasMapping;
}

/**
 * Resolves a list of service names to their canonical names, removing duplicates.
 *
 * Service aliases are resolved to their canonical service names based on the
 * alias mapping. Duplicates are removed to ensure each canonical service appears
 * only once in the result.
 *
 * @param serviceNames Array of service names (may include aliases)
 * @param aliasMapping Map from service name to canonical service name
 * @returns Array of unique canonical service names, sorted
 *
 * @example
 * ```typescript
 * const aliasMapping = new Map([
 *   ['historian', 'historian'],
 *   ['historian-secondary', 'historian'],
 * ]);
 * const resolved = resolveServiceAliases(['historian', 'historian-secondary'], aliasMapping);
 * console.log(resolved); // ['historian']
 * ```
 */
export function resolveServiceAliases(
  serviceNames: string[],
  aliasMapping: Map<string, string>
): string[] {
  const canonicalServices = new Set<string>();

  for (const serviceName of serviceNames) {
    const canonical = aliasMapping.get(serviceName);
    if (canonical) {
      canonicalServices.add(canonical);
    } else {
      // No mapping found - use service name as-is
      canonicalServices.add(serviceName);
    }
  }

  return Array.from(canonicalServices).sort();
}

/**
 * Normalizes a file path by joining context and filename.
 *
 * This helper function ensures consistent path formatting by:
 * 1. Removing trailing slashes from context
 * 2. Joining context and filename with a single forward slash
 *
 * @param context Build context directory path
 * @param filename Dockerfile name or relative path
 * @returns Normalized path joining context and filename
 *
 * @example
 * ```typescript
 * normalizePath('docker/app', 'Dockerfile'); // 'docker/app/Dockerfile'
 * normalizePath('docker/app/', 'Dockerfile'); // 'docker/app/Dockerfile'
 * normalizePath('docker', 'app/Dockerfile'); // 'docker/app/Dockerfile'
 * ```
 */
function normalizePath(context: string, filename: string): string {
  // Remove trailing slash from context
  const normalizedContext = context.replace(/\/$/, '');

  // Join with filename
  return `${normalizedContext}/${filename}`;
}

File: .github/scripts/detect-changes/lib/services.py
"""
Service discovery from docker-compose.yml.

Parses docker-compose.yml to find services with build configurations
and extracts their Dockerfile paths and contexts.
"""

import json
import subprocess
from pathlib import Path
from typing import Dict, List, Optional, Any


def discover_services_from_compose(compose_file: str) -> List[Dict[str, Any]]:
    """
    Discover services from docker-compose.yml using docker compose config.

    Executes 'docker compose --env-file example.env config --format json' to get
    the fully-resolved compose configuration, then filters for services with:
    - GHCR images (ghcr.io/groupsky/homy/*)
    - Build context defined

    Args:
        compose_file: Path to docker-compose.yml file

    Returns:
        List of service metadata dicts containing:
        - service_name: Name of the service
        - image: Full GHCR image path
        - build_context: Path to build directory
        - dockerfile_path: Resolved Dockerfile path
        - build_args: Build arguments (if present)

    Raises:
        RuntimeError: If docker compose command fails
        json.JSONDecodeError: If docker compose returns invalid JSON
    """
    # Get the directory containing the compose file
    compose_dir = Path(compose_file).parent

    # Execute docker compose config to get resolved configuration
    result = subprocess.run(
        [
            'docker', 'compose',
            '--env-file', 'example.env',
            '-f', compose_file,
            'config',
            '--format', 'json'
        ],
        cwd=compose_dir,
        capture_output=True,
        text=True
    )

    if result.returncode != 0:
        raise RuntimeError(f"docker compose config failed: {result.stderr}")

    # Parse JSON output
    config = json.loads(result.stdout)

    # Extract services section
    services_config = config.get('services', {})

    # Extract metadata for each service
    services = []
    for service_name, service_config in services_config.items():
        metadata = extract_service_metadata(service_name, service_config)
        if metadata is not None:
            services.append(metadata)

    # Filter to only GHCR-based services
    return filter_ghcr_services(services)


def extract_service_metadata(service_name: str, service_config: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Extract metadata from a service configuration.

    Args:
        service_name: Name of the service in docker-compose.yml
        service_config: Service configuration dict from docker compose config

    Returns:
        Service metadata dict or None if service has no valid build context

    The returned dict contains:
        - service_name: Name of the service
        - image: Full image path (may be None if not specified)
        - build_context: Path to build directory
        - dockerfile_path: Resolved Dockerfile path
        - build_args: Build arguments (if present)
    """
    # Get image (may not exist for some configs)
    image = service_config.get('image')

    # Get build configuration
    build = service_config.get('build')

    # If no build directive, return None
    if not build:
        return None

    # Handle string build context (shorthand notation)
    if isinstance(build, str):
        build_context = build.strip()
        dockerfile = 'Dockerfile'
        build_args = None
    # Handle dict build context (full notation)
    elif isinstance(build, dict):
        build_context = build.get('context')

        # If no context in dict, return None
        if not build_context:
            return None

        build_context = build_context.strip()
        dockerfile = build.get('dockerfile', 'Dockerfile').strip()
        build_args = build.get('args')
    else:
        # Invalid build type
        return None

    # Resolve dockerfile path
    dockerfile_path = str(Path(build_context) / dockerfile)

    # Build metadata dict
    metadata = {
        'service_name': service_name,
        'image': image,
        'build_context': build_context,
        'dockerfile_path': dockerfile_path,
    }

    # Add build args if present
    if build_args:
        metadata['build_args'] = build_args

    return metadata


def filter_ghcr_services(services: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Filter services to only include GHCR-based images.

    Filters for services with images starting with 'ghcr.io/groupsky/homy/'.

    Args:
        services: List of service metadata dicts

    Returns:
        Filtered list containing only GHCR-based services
    """
    ghcr_prefix = 'ghcr.io/groupsky/homy/'

    return [
        service for service in services
        if service.get('image') and service['image'].startswith(ghcr_prefix)
    ]
